### ISSUE TITLE
DVuploader for more number of files

### ISSUE THREAD

**Issue by linsherpa** [GlobalDataverseCommunityConsortium/dataverse-uploader#17]
I have a use case where:
The number of files for a dataset are in  thousands.

With dvuploader the upload speed decreases drastically with time as it checks the md5 of all files already uploaded in the dataset.

Are there any methods /ideas to bypass this check?

**Comment by qqmyers** [comment]
I think DVUploader only does this if you use the [-verify flag](https://github.com/GlobalDataverseCommunityConsortium/dataverse-uploader/wiki/DVUploader,-a-Command-line-Bulk-Uploader-for-Dataverse#optional-parameters).  It does always scan through the list of files in the dataset to get to the first yet-to-be-uploaded one, but it should only be calculating the hash with that parameter.  (That said, it's possible that there's a bug or this was different in older versions.). DVUploader does only do this once per run though, so one ~work-around would be to avoid using a -limit param and have it upload as many files as possible before a restart.

**Comment by linsherpa** [comment]
Thank you for the answer.

And sorry for miscommunication earlier.
Actually I am using the latest version of DV uploader 1.1.0
and the command goes like:
java -jar <DVUploader-(latestversion)> -key=<api key> -did=<dataset doi> -server=<server URL> <Maindirectoryname(path)> -recurse

It gives me a standard feedback as below:
```
PROCESSING(F): picture/Picture_A.jpg
                              Does not yet exist on server.
                              UPLOADED as: MD5:8b54f4fc16a57af76577a46eeeff355c                     
 ```
What I wanted to point out was that:                           
For me the Message "UPLOADED as MD5" takes longer time (in minutes) as the number of files already uploaded  in dataset increases.

So I was wondering if there are are any quick workaround.

**Comment by qqmyers** [comment]
Ah - the delay there is not in calculating the hash. DVUploader 1.1.0 uses the single file /add api call in Dataverse and, after every such call, Dataverse updates the dataset which includes reindexing all files. DVUploader waits for that process to finish to avoid sending the next file while the dataset is locked/being updated. There are a few things that can be done, depending on your configuration, to make the process faster now, and there's ongoing discussion/work about speeding things up. Work-arounds include:

- if your Dataverse supports it, use -directupload. This sends the files directly to S3 and avoids Dataverse making local copies and handling the upload itself. 
- if your Dataverse is using full-text-indexing, see if that can be turned off while uploading. Reindexing the dataset once after all files are loaded with full-text-indexing on is sufficient and faster
- if your Dataverse uses direct upload and is >=v5.6, use the as-yet-unreleased DVUploader which uses the new /addFiles API to add several files at once (currently hardcoded at ~10 I think - will probably be configurable in a release). This results in fewer updates of the dataset.

(Don't run multiple DVUploaders in parallel on the same dataset! Other copies will just find that the dataset is locked/run into errors due to parallel updates. (also DVUploader reads the dataset once to start and assumes that there are no changes while it is deciding which files to upload so the copies may try to upload the same file)).

**Comment by linsherpa** [comment]
thank you @qqmyers . 
I have tested the beta version of Dvuploader...its crazy fast.

Small comment: IT works only on bucket with "directupload=true" and with option -directupload on dvuploader.

Thanks again :)

**Comment by qqmyers** [comment]
~Fixed in v1.2.0 (now at beta3) for directupload (which is now the default). As I don't expect a fix/improvements for the non directupload case in DVUploader in the near term, I'm going to close this issue. (FWIW: Dataverse 5.14+ has some internal improvements that may speed the non-directupload case.)
